xDiffCollection/DiffCollection.swift

```diff
@@ -0,0 +1,111 @@
+//
+//  xDiffCollection.swift
+//  CollectionTest
+//
+//  Created by Esteban Garro on 2018-06-05.
+//  Copyright © 2018 aldo. All rights reserved.
+//
+
+import Foundation
+
+public struct DiffCollectionResult {
+    public var deleted : [IndexPath]
+    public var added   : [IndexPath]
+    public var updated : [IndexPath]
+}
+
+public struct DiffCollectionFilter<T:Hashable> {
+    public let name : String
+    public var filter:(T)->Bool
+    
+    public init(name: String, filter: @escaping (T)->Bool) {
+        self.name = name
+        self.filter = filter
+    }
+}
+
+public struct DiffCollection<T:Hashable> : CustomStringConvertible {
+    private var bins:[CollectionBin<T>] = []
+    private init(bins:[CollectionBin<T>]) {
+        self.bins = bins
+    }
+    
+    public var description: String {
+        var resp = ""
+        for bin in bins {
+            resp += "Bin\n"
+            resp += "\(bin)"
+        }
+        return resp
+    }
+    
+    public init(filters:[DiffCollectionFilter<T>]) {
+        for (idx, binFilter) in filters.enumerated() {
+            bins.append(CollectionBin(index: idx, filter: binFilter.filter, elements: []))
+        }
+    }
+    
+    public mutating func update(element:T) -> DiffCollectionResult {
+        var deletedItems : [IndexPath] = []
+        var addedItems : [IndexPath] = []
+        var updatedItems : [IndexPath] = []
+                
+        var resp : [CollectionBin<T>] = []
+        for (binIndex, bin) in bins.enumerated() {
```

Comment:
Like I told you during discussion. We could get the same result `DiffCollectionResult`  from bin.update. Then we could use flatMap to make the code more cleaner and smaller - what do you think?